Separate name from prefixes

ondenman opened the pull request:
This is a step towards scraping members who have left the legislature, along with their end dates. (Issue: https://github.com/everypolitician/everypolitician-data/issues/25279)

Some member names are displayed along with their titles.

![screen shot 2017-01-30 at 16 08 40](https://cloud.githubusercontent.com/assets/4107953/22430508/6c9607d4-e706-11e6-806a-b535c8d6a48f.png)

This PR separates the title from the name. It adds the field `:honorific_prefix`

ondenman committed:
Test that name does not include prefix

ondenman committed:
Remove prefix from name

ondenman committed:
Test presence of honorific_prefix field

ondenman committed:
Add :honorific_prefix

ondenman committed:
Test that honorific_prefix captures member title

ondenman committed:
Return titles from honorific_titles field

tmtmtmtm reviewed (changes requested):
It seems that the person being used in the test here doesn't currently exist in the data being scraped. 

I'd suggest you use as a test case someone who's actually being scraped so we can see this being used. A good set of test cases here would also include someone who has no prefix, and someone who has multiple prefixes. I'd also prefer to see a test case for a person who has the "(Ret’)" case, simply because it stands out as an edge-case (though as I'm assuming it never appears on its own, that's probably part of the "someone with multiple prefixes" case)

ondenman commented:
Addresses: https://github.com/everypolitician/everypolitician-data/issues/25279

ondenman committed:
Test member without honorific prefixes

ondenman commented:
I've added another subject so we can test a member without a prefix.

There are no members in the live data with prefixes. This change is being made in anticipation of adding ex members -- two of which have the 'Dr' prefix. So we are testing against the page of an ex member.

I have removed the (Ret') prefix in a fixup as this would not likely appear on its own and we're not testing for multiple prefixes.

tmtmtmtm reviewed (changes requested):
There still doesn't seem to be a test for people with multiple prefixes.

If the issue is that no-one actually _has_ multiple prefixes, then I'm not sure why we need such a big list of possible prefixes. If that's just a cut-n-paste from some other scraper that needed those, to cope with some future scenario when they actually elect a bishop or a prince, then don't do that. If we really want a generic list like this, then it should be encapsulated into a module somewhere.

I'd also suggest switching these to `to_h` driven tests like the other cases. That's what we're trying to move to everywhere, so adding new tests that _don't_ follow that pattern isn't really a good plan.

ondenman committed:
Test hashes instead of fields

tmtmtmtm commented on lib/member_page.rb:
```diff
@@ -36,6 +40,14 @@ class MemberPage < Scraped::HTML
+    %w(Dr)
+  end
+
+  def name_parts
```
There's really two different things going on in one method here — one is getting the text of the 'raw' name field — the other is processing that into parts. Those should really be separate methods.

tmtmtmtm commented on test/name_and_prefix_test.rb:
```diff
@@ -0,0 +1,49 @@
+    subject do
+      url = 'http://www.parlament.mt/borg-ian'
+      MemberPage.new(response: Scraped::Request.new(url: url).response)
+    end
```
There's a little too much duplication here. You should remove that in the same way that you have in other scrapers.

tmtmtmtm commented on test/name_and_prefix_test.rb:
```diff
@@ -0,0 +1,49 @@
+        image:             'http://www.parlament.mt/file.aspx?f=31734',
+        source:            'http://www.parlament.mt/borg-ian',
+        electoral_history: {}
+      ).must_equal true
```
that final `.must_equal true` seems a little redundant. More prosaically, it's different from how you're doing the equivalent test for Albert Fenech above, without explaining why.

tmtmtmtm commented on lib/member_page.rb:
```diff
@@ -36,6 +40,14 @@ class MemberPage < Scraped::HTML
+  end
+
+  def name_parts
+    box.css('h1').text.split(' - ').first.sub('Hon. ', '').sub(' MP', '').tidy.split(' ')
```
now that you're splitting the name up and putting it back together again anyway, it doesn't really make a lot of sense to manipulate the raw text like this. I'd suggest dropping that part of this, setting up separate lists, along the lines of:

```ruby
  def wanted_titles
    %w(Dr)
  end

  def unwanted_titles
    %w(Hon. MP)
  end

  def titles
    wanted_titles + unwanted_titles
  end
```

and then changing `honorifc_prefix` to only include parts in `wanted_titles`

tmtmtmtm committed:
fixup! Return titles from honorific_titles field

ondenman committed:
fixup! Test hashes instead of fields
